### ISSUE TITLE
Single Vault Encrypted value not decrypted in jinja2 pipeline

### ISSUE THREAD

**Issue by sivel** [ansible/ansible#24425]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
vault

##### ANSIBLE VERSION

```
devel
```

##### SUMMARY
A single vault encrypted value does not get decrypted in a jinja2 pipeline.

##### STEPS TO REPRODUCE

Don't worry the below vault data is just useless strings.

```yaml
---
- hosts: localhost
  gather_facts: false
  vars:
    the_password: !vault |
          $ANSIBLE_VAULT;1.1;AES256
          66653436353263336535316333313533666563386463623562396234313064643430323066626538
          3234393066313964393861613431313832306439383633340a653361636537303033313137613933
          33653333633139353736636332333237613761636430356665383839663164363739323337643333
          6466313434306462370a663539346162623836626165343830666536646631616165636438646639
          6134
  tasks:
    - command: 'echo {{ the_password|string|type_debug }}'
```

This outputs:

```
"stdout": "AnsibleVaultEncryptedUnicode"
```

This is a little bit of a contrived example, but what it shows, is that if you want to do something like:

```
- command: "some_command {{ the_password|quote }}"
```

You will end up with the `$ANSIBLE_VAULT;...` string instead of the decrypted value

##### EXPECTED RESULTS
```
```

##### ACTUAL RESULTS

```

```

**Comment by bcoca** [comment]
probably cause vault is only available from vars files, not from vars in play.

**Comment by sivel** [comment]
I was able to do something in the current form using the format filter to get it to work.

This comes from a user report on IRC.

I didn't want to lose this, so I did a little testing and filed it.

If you notice from the type_debug, it does indicate that it was understood properly.

**Comment by bcoca** [comment]
its the hackery in the base yaml constructor, but not something we had planned to use inline in plays (though I can see it's usefulness).

**Comment by sivel** [comment]
I'll double check, I don't think that is the problem though.  I just did it in this example for ease of use.  The OP didn't specifically state.

**Comment by sivel** [comment]
I've tested this with both vars_files and group_vars.  It's the same behavior.  It doesn't have anything to do with use of `vars`.

**Comment by sivel** [comment]
The only way I have gotten it to work is with:

```
'%s'|format(the_password)
```

That triggers the functionality in `AnsibleVaultEncryptedUnicode` to decrypt.  But passing through something like `|quote` or `|string` does not.

**Comment by sivel** [comment]
The following, on the surface seems to resolve this issue:

```diff
diff --git a/lib/ansible/parsing/yaml/objects.py b/lib/ansible/parsing/yaml/objects.py
index 41df711..c5fb4da 100644
--- a/lib/ansible/parsing/yaml/objects.py
+++ b/lib/ansible/parsing/yaml/objects.py
@@ -21,6 +21,8 @@ __metaclass__ = type
 
 import yaml
 
+from copy import deepcopy
+
 from ansible.module_utils.six import text_type
 from ansible.module_utils._text import to_bytes
 
@@ -110,9 +112,15 @@ class AnsibleVaultEncryptedUnicode(yaml.YAMLObject, AnsibleUnicode):
     def data(self, value):
         self._ciphertext = value
 
+    def __getattr__(self, name):
+        return self.data
+
     def __repr__(self):
         return repr(self.data)
 
+    def __deepcopy__(self, memo):
+        return deepcopy(self.data, memo)
+
     # Compare a regular str/text_type with the decrypted hypertext
     def __eq__(self, other):
         if self.vault:
```

I added `__deepcopy__` to prevent an exception, where I was returning `self.data` in `__getattr__`, but `deepcopy()` was looking for `__deepcopy__`.

**Comment by sivel** [comment]
Actually, it looks like the `__deepcopy__` is what does it.  The `__getattr__` isn't doing anything for me.

```diff
diff --git a/lib/ansible/parsing/yaml/objects.py b/lib/ansible/parsing/yaml/objects.py
index 41df711..5eb46bc 100644
--- a/lib/ansible/parsing/yaml/objects.py
+++ b/lib/ansible/parsing/yaml/objects.py
@@ -21,6 +21,8 @@ __metaclass__ = type
 
 import yaml
 
+from copy import deepcopy
+
 from ansible.module_utils.six import text_type
 from ansible.module_utils._text import to_bytes
 
@@ -113,6 +115,9 @@ class AnsibleVaultEncryptedUnicode(yaml.YAMLObject, AnsibleUnicode):
     def __repr__(self):
         return repr(self.data)
 
+    def __deepcopy__(self, memo):
+        return deepcopy(self.data, memo)
+
     # Compare a regular str/text_type with the decrypted hypertext
     def __eq__(self, other):
         if self.vault:
```

**Comment by bcoca** [comment]
totally misunderstood, reclasified

**Comment by pioto** [comment]
Thanks for doing the legwork here, @sivel -- I think I'm the IRC user you referred to.

Let me know if any other information is required to help resolve this.

**Comment by sivel** [comment]
Something else I noticed when I added `__getattr__` in the following manner:

```
def __getattr__(self, name):
    print(name)
    raise AttributeError(name)
```

I noticed the following calls to missing attributes:

* `__deepcopy__`
* `__getstate__`
* `__setstate__`

**Comment by alikins** [comment]
This looks like a dupe of #24175

Fixed in https://github.com/ansible/ansible/pull/24312 (merged in devel in 3840119bc75e74b721dd15828db128d1a8186309)

**Comment by alikins** [comment]
Though I think I like the idea of fixing with some combo of __getstate__/__deepcopy__ more than the slightly odd fix in https://github.com/ansible/ansible/pull/24312.  #24312 basic makes the default repr return self.data since templating does end up using the repr() but if that is only after try __getstate__ and/or __deepcopy__ I'd prefer adding those and letting the __repr__ be more useful.

**Comment by sivel** [comment]
@alikins I tried with your fix that you put in in #24312 and it does not resolve this issue.  That was one of the first things I suspected, is that this would be fixed by your change.  Unfortunately it isn't.

I imagine that is due to the pipelining of jinja2 filters, and what each individual filter does to the variable.

**Comment by sivel** [comment]
when piping to our `quote` filter, it gets passed directly into `pipes.quote` which doesn't seem to call `__str__`, `__unicode__`, or `__repr__`.  As such, it never decrypts the data.

**Comment by alikins** [comment]
@sivel yup, I assumed this was fixed in devel but I can reproduce the error with devel as well.

(removed the duplicate of)

**Comment by alikins** [comment]
@sivel 

> when piping to our quote filter, it gets passed directly into pipes.quote which doesn't seem to call __str__, __unicode__, or __repr__.

Interesting. That would explain where deepcopy/getstate come into play...

**Comment by pioto** [comment]
The fix on that other issue doesn't seem to be sufficient.

I've tried now with the current `stable-2.3` version of the code:

```
ansible-playbook 2.3.1.0 (stable-2.3 5f0be9dcd0) last updated 2017/05/10 10:38:18 (GMT -400)
  config file = /Users/mkelly/git/puppetless-servers-playbook/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Apr 23 2017, 16:50:35) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

And, I still see this issue.

Sample playbook:

```
---
- name: Configure IPMI
  hosts: nas
  tasks:
      - name: Set IPMI Password
        become: yes
        command: ipmitool user set password 2 {{ servers_password | quote }}
```

My command being run ends up as:

```
"cmd": ["ipmitool", "user", "set", "password", "2", "$ANSIBLE_VAULT;1.1;AES256\n39...\n"]
```

**Comment by alikins** [comment]
I think I see how to make AnsibleVaultEncryptedUnicode work for 'quote'...

pipes.quote() actually iterates over a string object to get each char and never uses the str()/unicode().
At the moment:

``` python

for i in some_ansible_vault_encrypted_unicode:
    print(i)

```

The value of it will be the chars in the ciphertext, not the plaintext.

Adding an __iter__ seems to fix it

``` python
   def __iter__(self):
         return iter(self.data)
```

**Comment by kaikuchn** [comment]
Just stumpled on this while trying to render a multiline secret into a yml file where I need to adhere to a specific indentation. I therefore wanted to pipe the decrypted secret through the `indent` filter only to discover that it would not longer decrypt then.

I understand that it may not even be desirable to allow secrets to be passed through filters.
So in an attempt to spare others the confusion, could you add a hint in the docs that one cannot use filters with vault encrypted variables? Thank you!

**Comment by jess-sol** [comment]
Ditto to @kaikuchn, ran into this trying to run indent on a secret.  The `'%s'|format()` trick did it for me though.  Maybe be good to document this

**Comment by Sodki** [comment]
We use a different workaround, by setting a fact to its own value before using it:

```
- set_fact:
  problematic_variable: "{{ problematic_variable }}"
```

This way the variable gets set to its decrypted value. The good thing about this approach is that you only need to add lines, not change the lines that you already have. Once the issue is fixed, the extra lines are gone.

**Comment by frolundo** [comment]
Appears to be same issue when using to_nice_yaml filter in .j2 template, like:
{{ system_creds | to_nice_yaml }}

Workaround with set_fact + no_log (to hide sensitive information) works fine, however, appears to be an issue with filters.

**Comment by stpierre** [comment]
The `__deepcopy__` approach seems simple, and fixes everything; why hasn't that been adopted yet? I've read through this and the other related issues and I can't find anywhere that specifies what the drawback is. We're suffering with this (mis)feature right now and it'd be really nice to know that it'll be fixed in 2.5.

**Comment by kaikuchn** [comment]
I don't remember where I read this (it is 3 months ago that I ran into this issue), but I think someone said that allowing secrets in filters would keep them in memory unencrypted which is an security issue. More so since the general assumption is that these parts of the memory do not contain sensitive data and thus might be freely leaked elsewhere.
However, I'm highly uncertain whether or not I remember this correctly...

**Comment by c4milo** [comment]
Any update on this?

**Comment by dbstraffin** [comment]
I have the same issue with the `to_nice_yaml` filter.

I worked around it with an intermediate var:
```yaml
creds_output: "{{ creds }}"
```
in template:
```
{{ creds_output | to_nice_yaml }}
```

**Comment by vide** [comment]
@Sodki Thanks for the workaround, it works perfectly. BTW this bug is still present with Ansible 2.6.0

**Comment by smithfred** [comment]
> I think someone said that allowing secrets in filters would keep them in memory unencrypted which is an security issue. More so since the general assumption is that these parts of the memory do not contain sensitive data and thus might be freely leaked elsewhere.

Can someone confirm/refute this?

If it's true, then what is the current "recommended" way to use e.g. the `password_hash` filter? That seems like an obvious case for wanting to store data in a vault file/var and then pass it through a filter.

**Comment by linnil1** [comment]
It still not work for `password_hash`

It need to add `string` before using `password_hash`

``` yaml
    - debug:
        msg: "{{ pw | string | password_hash('sha512') }}"
      vars: 
        pw: !vault |
          $ANSIBLE_VAULT;1.1;AES256
          626235
```

**Comment by juanvalino** [comment]
I have the same problem with 2.7.1. The workaround is to add `| string` before using `password_hash`.

```
ansible 2.7.1
  config file = /home/jvg/.ansible.cfg
  configured module search path = [u'/home/jvg/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jvg/.local/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:24:06) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

**Comment by kladiv** [comment]
+1

**Comment by teadur** [comment]
+1

**Comment by sivel** [comment]
After looking through what jinja2 is doing, effectively to exercise raise an undefined error, we should probably mimic that behavior.

Make sure that we explicitly exercise the code that results in decrypting the vault.

Such as forcing the `quote` filter, to explicitly call `to_text` on the data.

Instead of assuming that values are certain types, we should explicitly evaluate them, so that the special dunder methods of `AnsibleVaultEncryptedUnicode` are called.

**Comment by bcoca** [comment]
fixed in #57504

**Comment by akamac** [comment]
The workaround for dict with encrypted items, like
```
config:
  option: plaintext
  secure_option: !vault |
    $ANSIBLE_VAULT;1.1;AES256
    32336237613836383838656464316166346331393165333934396539346431653130623638656137
```
is to use `{{ config | string | from_yaml }}`.
Then you get unencrypted dict.

**Comment by Poil** [comment]
> string

Not working for me, it gave me an unicode dict

```
'{u''signalFxAccessToken'': u''xxxxxxxxxxx\n'', u''monitors'': [{u''type'':
    u''cpu''}, {u''type'': u''filesystems''}, {u''type'': u''disk-io''}, {u''type'':
    u''net-io''}, {u''type'': u''load''}, {u''type'': u''memory''}, {u''type'': u''vmem''},
    {u''type'': u''host-metadata''}, {u''type'': u''processlist''}]}'
```

**Comment by marwatk** [comment]
We ran into this issue without trying to use vaults in the play. Given an encrypted vault in a `group_vars/all.yml` like this:

```
test_inline_encryption: !vault |
  <Encrypted value>
```

These were the results we found trying to use it in various ways:
```
- hosts: localhost
  gather_facts: no
  connection: local
  tasks:

  - name: Result: Not decrypted
    debug:
      msg: "{{ vars_subkey | to_yaml | from_yaml }}"
    vars:
      vars_subkey: "{{ {'test_inline_encryption': test_inline_encryption } }}"

  - name: Result: Not decrypted
    set_fact:
      subkey: "{{ {'test_inline_encryption': lookup('vars', 'test_inline_encryption') } }}"
  - debug:
      msg: "{{ lookup('vars', 'subkey') }}"

  - name: Result: Not decrypted
    set_fact:
      subkey_nl: "{{ {'test_inline_encryption': test_inline_encryption } }}"
  - debug:
      msg: "{{ subkey_nl }}"

  - name: Result: Properly decrypted
    set_fact:
      fact_nl: "{{ test_inline_encryption }}"
  - debug:
      msg: "{{ fact_nl }}"

  - name: Result: Properly decrypted
    set_fact:
      a_fact: "{{ lookup('vars', 'test_inline_encryption') }}"
  - debug:
      msg: "{{ lookup('vars', 'a_fact') }}"

  - name: Result: Properly decrypted
    debug:
      msg: "{{ lookup('vars', 'test_inline_encryption') }}"
```

**Comment by feketegy** [comment]
If you want to use a vault encrypted variable in the user module, where the module requires the user password to be passed as a hash this is how you do it:

```
- name: Create user
  user:
    name: "testuser"
    password: "{{ user_vaulted_pass | string | password_hash('sha512') }}"
```

Basically, the vaulted password will be decrypted, converted to a string and re-hashed using sha512 and that will be passed as the user password for the user module.
